File: src/0535_encode_and_decode_tinyurl.js
class Codec {
    constructor() {
        this.encodeMap = new Map();
        this.decodeMap = new Map();
        this.base = 'http://tinyurl.com/';
    }
    encode(longUrl) {
        if (!this.encodeMap.has(longUrl)) {
            const shortUrl = this.base + String(this.encodeMap.size + 1);
            this.encodeMap.set(longUrl, shortUrl);
            this.decodeMap.set(shortUrl, longUrl);
        }
        return this.encodeMap.get(longUrl);
    }
    decode(shortUrl) {
        return this.decodeMap.get(shortUrl);
    }
}

module.exports = { Codec };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    let codec = new Codec();
    let url = "https://leetcode.com/problems/design-tinyurl";
    let short_url = codec.encode(url);
    assert.deepStrictEqual(short_url, "http://tinyurl.com/1");
    assert.deepStrictEqual(codec.decode(short_url), url);
    assert.deepStrictEqual(codec.encode(url), short_url);
}

File: src/0535_encode_and_decode_tinyurl.rs
use std::collections::HashMap;

struct Codec {
    encode_map: HashMap<String, String>,
    decode_map: HashMap<String, String>,
    base: String,
}

impl Codec {
    fn new() -> Self {
        Self {
            encode_map: HashMap::new(),
            decode_map: HashMap::new(),
            base: "http://tinyurl.com/".to_string(),
        }
    }

    fn encode(&mut self, long_url: String) -> String {
        if !self.encode_map.contains_key(&long_url) {
            let short_url = format!("{}{}", self.base, self.encode_map.len() + 1);
            self.encode_map.insert(long_url.clone(), short_url.clone());
            self.decode_map.insert(short_url, long_url.clone());
        }

        self.encode_map[&long_url].clone()
    }

    fn decode(&self, short_url: String) -> String {
        self.decode_map[&short_url].clone()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        let mut codec = Codec::new();
        let url = String::from("https://leetcode.com/problems/design-tinyurl");
        let short_url = codec.encode(url.clone());
        assert_eq!(short_url, String::from("http://tinyurl.com/1"));
        assert_eq!(codec.decode(short_url.clone()), url);
        assert_eq!(codec.encode(url.clone()), short_url);
    }
}
